### ISSUE TITLE
Improve output of `Object:throw` for non-exception objects

### ISSUE THREAD

**Issue by jamshark70** [supercollider/supercollider#3612]
We can `throw` any object. However, the printed output is misleading and has led to at least one case where a developer wanted to throw an error, but threw a string, and it didn't work as expected (#3357).

**A.** Throwing an object masquerades as an error -- `Object:reportError` fakes an error string and dumps the stack.

```
Error("abc").throw;

"abc".throw;
```

There's essentially no difference in the output between them (except that the Error posts the "preceding error" line, while the string alone does not). From this, users may draw the conclusion that `Error(...)` is optional. It isn't.

So, we should change the `Object:reportError` output to make it clear that the user has thrown something that isn't an exception, and it isn't being handled -- which is user error.

**B.** Throwing an object is legitimate in a `try` block, where it functions as an early exit mechanism.

```
(
x = try {
	10.do { |i|
		if(i == 5) { "yes yes".throw };
	};
	"no no"
} { |error| "got error".postln; error.dump };

(x ++ " 123").postln;
)

Posts:
yes yes 123
-> yes yes 123
```

In this usage, the error handler is not called at all. If we want to allow this usage for early exit, that's correct (emphasizing that, if you need to throw an error, you must throw an `Error`).

This also works for `protect`, in which case it *does* call the unwind function -- I think this is correct as well -- but the `|error|` argument is nil, and execution continues after the `protect` block. So you can early exit, rely on the unwind function to close a file for instance, and keep going (and the unwind function can differentiate between early exit and a fatal error based on its argument).

I think `try` and `protect` are fine here -- but, it's too confusing that you can throw a non-error and it looks almost exactly like throwing an error.

**Comment by telephon** [comment]
I would like to discuss whether we want to support two or three levels. The three levels are:

1. object thrown
2. exception thrown (`isException == true`)
3. error thrown

`try` and `protect` should behave the same in this respect. But the could either include 2. and 3., or also include 1.

The only reason to have a separate layer (1) would be that it can be a valid way to control program flow to throw and `catch`, without that having to be an exception.

So if we want to support 3 levels, it would make sense to add a method like `catch` which is like `try` but for all objects thrown

Both ways seem reasonable, however.

**Comment by telephon** [comment]
@muellmusik linked this: https://yosefk.com/c++fqa/exceptions.html#fqa-17.6

**Comment by jamshark70** [comment]
At present, as far as I can see, the only reason to have an `Exception` class separate from `Error` is in case somebody, down the road, needs the distinction. But currently, there is no functional difference. `Exception("what").throw` prints `EXCEPTION: what` instead of `ERROR: what`. That's it.

So 2 vs 3 is presently not a useful distinction. https://www.quora.com/How-do-I-tell-a-person-who-is-not-good-in-programming-the-difference-between-error-and-exception -- If we accept this explanation, then pretty much everything sclang calls an error is really an exception.

In the other issue, I had asked why we could throw any object but only Exceptions go through error handling. Then I poked around and answered that question -- early exit. We don't "need" this approach, necessarily, because we have `block`, but it's probably nice to keep it in case you need both early exit and error handling:

```
try {
	block { |break|
		loop with 'break.value(return)'
	}
} { |error| ... }

// vs

try {
	loop with 'return.throw'
} { |error| ... }
```

**Comment by telephon** [comment]
Yes, that makes sense: `2` and `3` are different only in the potential behavior of the object thrown, not in the way they are passed through the system. The message `isException` is the crucial differentiation point here.

And indeed for `1` I had thought of it being part of a control structure. What it does and `block` doesn't do is to percolate up through the thread structure. This is probably some equivalent to what in other languages is mentioned as "stack unwinding". As far as I see, we don't need the latter. But when we are in a routine that has been forked off from another one, it is useful to notify the whole chain of some, well "object thrown". I had implemented that at some stage, and it is used in `Pprotect`.

```supercollider
a = Task {  1.wait; a.isPlaying.postln; fork { 1.wait; "stop".throw } }.play;
// wait a little
a.isPlaying;
```

The question is whether we want to bring `anObject.throw` closer to `anException.throw` or keep it distinct.

So either, 

A) `anObject.throw` should also be caught by `try`
B) `anObject.throw` should not be caught by `try`or `protect`, but by a different method `catch`.

But currently, the wording in the whole system suggests that the intention is to treat only errors. So `B` would be something to go through very carefully.
